framework/src/meshgenerators/XYZDelaunayGenerator.C

```diff
@@ -119,13 +131,76 @@ XYZDelaunayGenerator::generate()
+
   // Get ready to triangulate its boundary
   libMesh::NetGenMeshInterface ngint(*mesh);
 
   ngint.smooth_after_generating() = _smooth_tri;
 
   ngint.desired_volume() = _desired_volume;
 
+  // The hole meshes will be used for hole boundary identification and optionally for stitching.
+  // if a hole mesh contains 3D volume elements but has non-TRI3 surface side elements, it cannot be
+  // used directly for stitching. But it can be converted into an all-TET4 mesh to support hole
+  // boundary identification
+  for (auto hole_i : index_range(_hole_ptrs))
+  {
+    UnstructuredMesh & hole_mesh = dynamic_cast<UnstructuredMesh &>(**_hole_ptrs[hole_i]);
+    libMesh::MeshSerializer serial_hole(hole_mesh);
+    // Check the dimension of the hole mesh
+    // We do not need to worry about element order here as libMesh checks it
+    std::set<ElemType> hole_elem_types;
+    std::set<unsigned short> hole_elem_dims;
+    for (auto elem : hole_mesh.element_ptr_range())
+    {
+      hole_elem_dims.emplace(elem->dim());
+
+      // For a 3D element, we need to check the surface side element type instead of the element
+      // type
+      if (elem->dim() == 3)
+        for (auto s : make_range(elem->n_sides()))
+        {
+          if (!elem->neighbor_ptr(s))
+            hole_elem_types.emplace(elem->side_ptr(s)->type());
+        }
+      // For a non-3D element, we just need to record the element type
+      else
+        hole_elem_types.emplace(elem->type());
+    }
+    if (hole_elem_dims.size() != 1 || *hole_elem_dims.begin() < 2)
+      paramError(
+          "holes",
+          "All elements in a hole mesh must have the same dimension that is either 2D or 3D.");
+    else if (*hole_elem_dims.begin() == 3)
+    {
+      // For 3D meshes, if there are non-TRI3 surface side elements
+      // (1) if no stitching is needed, we can just convert the whole mesh into TET to facilitate
+      // boundary identification (2) if stitching is needed, we can still convert and stitch, but
+      // that would modify the input hole mesh
+      if (*hole_elem_types.begin() != ElemType::TRI3 || hole_elem_types.size() > 1)
+      {
+        if (_stitch_holes.size() && _stitch_holes[hole_i] && !_convert_holes_for_stitching)
+          paramError("holes",
+                     "3D hole meshes with non-TRI3 surface elements cannot be stitched without "
+                     "converting them to TET4. Consider setting convert_holes_for_stitching=true.");
+        else
+          MeshTools::Modification::all_tri(**_hole_ptrs[hole_i]);
+      }
+    }
+    else // if (*hole_elem_dims.begin() == 2)
+    {
+      // There is no point to stitch a 2D hole mesh, so we throw an error if this is attempted
+      if (_stitch_holes.size() && _stitch_holes[hole_i])
+        paramError("holes",
+                   "the hole mesh with index " + std::to_string(hole_i) +
+                       " is a 2D mesh, for which stitching onto a 3D mesh does not make sense.");
+    }
```

Comment:
what about if the 2D surface has a mix of tris and quads?

Reply:
I see you resolved this, what am I missing?

Reply:
Weird, I answered this one and another but it disappeared...

Anyway, the libmesh's netgen interface converts quads to tris before using it in tetragonization. However, it does not covert the original hole meshes. Therefore, if stitching is needed, we need to convert the mesh here.